node/src/components/linear_chain_sync.rs

```diff
@@ -364,7 +365,10 @@ impl<I: Clone + PartialEq + 'static> LinearChainSync<I> {
 
     // Returns whether we've just downloaded a switch block of a currently active era.
     fn is_currently_active_era(&self, block: &Block) -> bool {
-        block.header().timestamp().elapsed() < self.shortest_era
+        let last_switch_block_height = self.state.last_switch_block_height().unwrap_or(0);
+        let past_blocks_in_this_era = block.height().saturating_sub(last_switch_block_height);
+        block.header().timestamp().elapsed() + self.min_round_length * past_blocks_in_this_era
+            < self.shortest_era
```

Comment:
I'm still confused. To my mind, this should return `true` if either:
* the _switch block_'s timestamp plus the minimum era length is in the future, and
* the new `block`'s timestamp plus `min_round_length` times the number of _remaining_ blocks in this era is in the future.

Is this really equivalent? `block.header().timestamp().elapsed() + self.min_round_length * past_blocks_in_this_era` looks like an _estimate_ of the switch block's timestamp, even though we could know it. I'd rather save that timestamp (or the switch block header) in addition to `last_switch_block_height`.

I also don't think there's a reason anymore to compute `shortest_era` in advance, since we need to use minimum round length and minimum era length and height in this method explicitly anyway.

Reply:
This is equivalent to the second point. `block.header().timestamp().elapsed()` translates to `now - block.header().timestamp()`. So we have:
`now - block.header().timestamp() + self.min_round_length * past_blocks_in_this_era < self.shortest_era`
`now < block.header().timestamp() + self.shortest_era - self.min_round_length * past_blocks_in_this_era`

`self.shortest_era - self.min_round_length * past_blocks_in_this_era` is the number of remaining blocks times the `min_round_length`. So this is really your second point, but expressed in terms of what we have available, and avoiding the fact that `TimeDiff`s can't be negative (so minimizing subtractions).

Actually, the first point is also taken care of when we handle the switch block. We will then check whether time elapsed from its timestamp is less than the shortest era, and if yes, we will return true.

So yes, this is equivalent, though maybe not the most intuitive :grimacing:

Reply:
Right, makes sense! Thank you for the explanation! A comment in the code would be helpful, though.